### PR TITLE
Add checkups for installation troubleshooting + general Windows troubleshooting

### PR DESCRIPTION
**Pull request by RebeccaMahany**
macOS:
* Grabs installation logs

Windows:
* Get all running services
* Get status and config for Launcher service
* Get recent service manager event logs for Launcher service
* Generate systempowerreport

### CODE REVIEW

File: pkg/debug/checkups/install.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"io"
+	"os"
+	"runtime"
+)
+
+type installCheckup struct {
+}
+
+func (i *installCheckup) Name() string {
+	return "Installation"
```
Comment (directionless) [inline pkg/debug/checkups/install.go:16]:
```suggestion
	return "Package Install Logs"
```

File: pkg/debug/checkups/install.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (i *installCheckup) Run(ctx context.Context, extraWriter io.Writer) error {
+	extraZip := zip.NewWriter(extraWriter)
+	defer extraZip.Close()
+
+	if err := gatherInstallationLogs(extraZip); err != nil {
+		return fmt.Errorf("gathering installation logss: %w", err)
```
Comment (directionless) [inline pkg/debug/checkups/install.go:24]:
```suggestion
		return fmt.Errorf("gathering installation logs: %w", err)
```

File: pkg/debug/checkups/install.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+func (i *installCheckup) Data() any {
+	return nil
+}
+
+func gatherInstallationLogs(z *zip.Writer) error {
+	if runtime.GOOS == "windows" || runtime.GOOS == "linux" {
+		return nil
+	}
+
+	out, err := z.Create("install.log")
```
Comment (directionless) [inline pkg/debug/checkups/install.go:52]:
```suggestion
	out, err := z.Create("macos-var-log-install.log")
```

File: pkg/debug/checkups/install.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		return fmt.Errorf("creating install.log in zip: %w", err)
+	}
+
+	installLog, err := os.Open("/var/log/install.log")
+	if err != nil {
+		return fmt.Errorf("opening /var/log/install.log: %w", err)
+	}
+	defer installLog.Close()
+
+	installLogRaw, err := io.ReadAll(installLog)
```
Comment (directionless) [inline pkg/debug/checkups/install.go:63]: Maybe copy instead of readall

File: pkg/debug/checkups/services.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+const kolideSvcName = "LauncherKolideK2Svc"
+
+type servicesCheckup struct {
+	data         map[string]any
+	serviceState svc.State
+}
+
+func (s *servicesCheckup) Name() string {
+	return "Service Report"
+}
```
Comment (directionless) [inline pkg/debug/checkups/services.go:23]:
```suggestion
func (s *servicesCheckup) Name() string {
	if runtime.GOOS != "windows" {
		return ""
	}
	return "Service Report"
}
```

Comment (directionless) [reply pkg/debug/checkups/services.go:23]: Though given that svc only exist on windows, we're going to need to build constrain this anyhow :|

File: pkg/debug/checkups/services.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+	if err != nil {
+		return fmt.Errorf("opening service: %w", err)
+	}
+	defer serviceHandle.Close()
+
+	s.data["service_name"] = kolideSvcName
+
+	// Get status
+	if status, err := serviceHandle.Query(); err != nil {
+		s.data["state"] = fmt.Sprintf("error: %v", err)
```
Comment (directionless) [inline pkg/debug/checkups/services.go:48]: Could also populate this into the Summary

File: pkg/debug/checkups/services.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	} else {
+		// Map state (uint32) to human-readable statuses
+		s.serviceState = status.State
+		s.data["state_raw"] = status.State
+		s.data["state"] = stateHumanReadable(status.State)
+	}
+
+	// Get config
+	if cfg, err := serviceHandle.Config(); err != nil {
+		s.data["config"] = fmt.Sprintf("error: %v", err)
```
Comment (directionless) [inline pkg/debug/checkups/services.go:58]: Same

Comment (RebeccaMahany) [reply pkg/debug/checkups/services.go]: i'm not including config info in the summary right now, so i'm not sure how to fit this in gracefully. is there config info that you think should go in the summary?

Comment (directionless) [reply pkg/debug/checkups/services.go]: I wasn't thinking about any info, but failing to parse it seems like it's probably an error condition worth noting.

File: pkg/debug/checkups/power_windows.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
+//go:build windows
+// +build windows
```
Comment (directionless) [inline pkg/debug/checkups/power_windows.go:2]: This is fine, but we don't need to build tag -- can switch in the `Name()`. Mostly depends what feels easier

File: pkg/debug/checkups/power_windows.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+func (p *powerCheckup) Name() string {
+	return "Power Report"
+}
+
+func (p *powerCheckup) Run(ctx context.Context, extraWriter io.Writer) error {
+	// Create a temporary file for powercfg to write its output to
+	tmpFilePath := agent.TempPath("launcher-checkup-spr.html")
+	defer func() {
+		os.Remove(tmpFilePath)
+	}()
```
Comment (directionless) [inline pkg/debug/checkups/power_windows.go:28]:
```suggestion
	defer os.Remove(tmpFilePath)
```

maybe?

File: pkg/debug/checkups/power_windows.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+	// See: https://learn.microsoft.com/en-us/windows-hardware/design/device-experiences/powercfg-command-line-options#option_systempowerreport
+	powerCfgCmd := exec.CommandContext(ctx, "powercfg.exe", "/systempowerreport", "/output", tmpFilePath)
+	powerCfgCmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true} // prevents spawning window
```
Comment (directionless) [inline pkg/debug/checkups/power_windows.go:32]: nice

File: pkg/debug/checkups/power_windows.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	"os/exec"
+	"syscall"
+
+	"github.com/kolide/launcher/pkg/agent"
+)
+
+type powerCheckup struct{}
+
+func (p *powerCheckup) Name() string {
+	return "Power Report"
```
Comment (directionless) [inline pkg/debug/checkups/power_windows.go:20]: If this is destined to only be windows, I think we should include that in the name.

File: pkg/debug/checkups/services_windows.go [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+	default:
+		return fmt.Sprintf("unknown: %d", serviceType)
+	}
+}
+
+func startTypeHumanReadable(startType uint32) string {
+	// For mapping, see https://pkg.go.dev/golang.org/x/sys/windows/svc/mgr#pkg-constants
+	// and https://pkg.go.dev/golang.org/x/sys@v0.11.0/windows#SERVICE_BOOT_START
+	switch startType {
+	case 0:
```
Comment (directionless) [inline pkg/debug/checkups/services_windows.go:160]: These are constants in the windows package, should we use them? `case windows.SERVICE_BOOT_START` or something?

File: pkg/debug/checkups/services_windows.go [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+
+func gatherServiceManagerEventLogs(ctx context.Context, z *zip.Writer) error {
+	cmdletArgs := []string{
+		"Get-EventLog",
+		"-Newest", "50",
+		"-LogName", "System",
+		"-Source", "\"Service Control Manager\"",
+		"-Message", fmt.Sprintf("*%s*", kolideSvcName),
+		"|",
+		"Format-Table", "-Wrap", "-AutoSize", // ensure output doesn't get truncated
```
Comment (directionless) [inline pkg/debug/checkups/services_windows.go:237]: Does this actually expand the messages? I know I've had trouble with that before.

There may also be a to json tool

File: pkg/debug/checkups/services_windows.go [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+	}
+
+	getEventLogCmd := exec.CommandContext(ctx, "powershell.exe", cmdletArgs...)
+	getEventLogCmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true} // prevents spawning window
+	out, err := getEventLogCmd.CombinedOutput()
```
Comment (directionless) [inline pkg/debug/checkups/services_windows.go:242]: You might be able to create the writer, and then set `getEventLogCmd.Stdout` and getEventLogCmd.Stderr` to the writer directly.

File: pkg/debug/checkups/power_windows.go (hunk 8 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/power_windows.go:2]: We do have to use the build tag for this, unfortunately -- before I separated them out and added the tag: https://github.com/kolide/launcher/actions/runs/5813681186/job/15761773747

File: pkg/debug/checkups/power_windows.go (hunk 11 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/power_windows.go:20]: I thought maybe we'd want to pull something similar for macos/linux in the future? And we probably don't need to include the OS for the device in the checkup name

File: pkg/debug/checkups/services_windows.go (hunk 13 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/services_windows.go:237]: It does, yes

File: pkg/debug/checkups/power_windows.go (hunk 9 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/power_windows.go:28]: Oh yup, no good reason to do it the way I have it now, I'll update

File: pkg/debug/checkups/services_windows.go (hunk 12 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/services_windows.go:160]: Oh yup that would improve readability a lot -- thank you, I'll update

File: pkg/debug/checkups/services_windows.go (hunk 14 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/services_windows.go]: Yup, that works! Updated

File: pkg/debug/checkups/services.go (hunk 7 above)
Comment (RebeccaMahany) [reply pkg/debug/checkups/services.go]: Added

**Review by directionless — APPROVED** [review]
Woo! LGTM

**Pull request merged**